File: crates/juicefront/ui/src/lib/legal.ts
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

let _publicDir: string | null = null;

function getPublicDir(): string {
  if (_publicDir) return _publicDir;
  const here = path.dirname(fileURLToPath(import.meta.url));
  let dir = here;
  while (dir !== path.dirname(dir)) {
    if (fs.existsSync(path.join(dir, "public"))) {
      _publicDir = path.join(dir, "public");
      return _publicDir;
    }
    dir = path.dirname(dir);
  }
  _publicDir = path.resolve("public");
  return _publicDir;
}

export function readPublicFile(name: string): string {
  const ext = path.extname(name);
  const base = name.slice(0, -ext.length);
  const here = path.dirname(fileURLToPath(import.meta.url));
  const candidates = [
    path.join(getPublicDir(), name),
    path.resolve("client", name),
    path.join(here, "../client", name),
    path.join(here, "../public", name),
    path.join(getPublicDir(), `${base}.example${ext}`),
  ];
  for (const p of candidates) {
    try {
      return fs.readFileSync(p, "utf-8");
    } catch {}
  }
  return "";
}

function escapeHtml(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#39;");
}

function inline(text: string): string {
  return escapeHtml(text)
    .replace(/\*\*(.+?)\*\*/g, "<strong>$1</strong>")
    .replace(/\*(.+?)\*/g, "<em>$1</em>");
}

export function renderTxtToHtml(raw: string): string {
  const lines = raw.replace(/\r\n/g, "\n").split("\n");
  const out: string[] = [];
  let inList = false;
  let para: string[] = [];

  const closePara = () => {
    if (para.length) {
      out.push(`<p>${inline(para.join("<br>"))}</p>`);
      para = [];
    }
  };
  const closeList = () => {
    if (inList) {
      out.push("</ul>");
      inList = false;
    }
  };

  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed) {
      closeList();
      closePara();
      continue;
    }
    const heading = trimmed.match(/^(#{1,3})\s+(.*)$/);
    if (heading) {
      closeList();
      closePara();
      const level = Math.min(heading[1].length + 1, 3);
      out.push(`<h${level}>${inline(heading[2])}</h${level}>`);
      continue;
    }
    if (/^---+\s*$/.test(trimmed)) {
      closeList();
      closePara();
      out.push("<hr>");
      continue;
    }
    const banner = trimmed.match(/^---+\s+(.+?)\s+---+$/);
    if (banner) {
      closeList();
      closePara();
      out.push(`<h2>${inline(banner[1])}</h2>`);
      continue;
    }
    const item = trimmed.match(/^\s*[-*]\s+(.*)$/);
    if (item) {
      closePara();
      if (!inList) {
        out.push("<ul>");
        inList = true;
      }
      out.push(`<li>${inline(item[1])}</li>`);
      continue;
    }
    closeList();
    para.push(trimmed);
  }
  closeList();
  closePara();
  return out.join("\n");
}

File: crates/juicefront/src/ssr/legal.rs
pub const TERMS_TXT: &str = include_str!("../../static/public/terms-of-service.txt");
pub const PRIVACY_TXT: &str = include_str!("../../static/public/privacy-policy.txt");

pub static TERMS_HTML: std::sync::LazyLock<String> =
    std::sync::LazyLock::new(|| render_txt_to_html(TERMS_TXT));
pub static PRIVACY_HTML: std::sync::LazyLock<String> =
    std::sync::LazyLock::new(|| render_txt_to_html(PRIVACY_TXT));

pub fn render_txt_to_html(raw: &str) -> String {
    let mut out: Vec<String> = Vec::new();
    let mut in_list = false;
    let mut para: Vec<String> = Vec::new();

    let close_para = |out: &mut Vec<String>, para: &mut Vec<String>| {
        if !para.is_empty() {
            out.push(format!("<p>{}</p>", inline(&para.join("<br>"))));
            para.clear();
        }
    };
    let close_list = |out: &mut Vec<String>, in_list: &mut bool| {
        if *in_list {
            out.push("</ul>".to_owned());
            *in_list = false;
        }
    };

    for line in raw.replace("\r\n", "\n").split('\n') {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            close_list(&mut out, &mut in_list);
            close_para(&mut out, &mut para);
            continue;
        }
        if let Some(heading) = parse_heading(trimmed) {
            close_list(&mut out, &mut in_list);
            close_para(&mut out, &mut para);
            out.push(heading);
            continue;
        }
        if trimmed.chars().all(|ch| ch == '-') && trimmed.len() >= 3 {
            close_list(&mut out, &mut in_list);
            close_para(&mut out, &mut para);
            out.push("<hr>".to_owned());
            continue;
        }
        if let Some(banner) = parse_banner(trimmed) {
            close_list(&mut out, &mut in_list);
            close_para(&mut out, &mut para);
            out.push(banner);
            continue;
        }
        if let Some(item) = parse_item(trimmed) {
            close_para(&mut out, &mut para);
            if !in_list {
                out.push("<ul>".to_owned());
                in_list = true;
            }
            out.push(format!("<li>{}</li>", inline(item)));
            continue;
        }
        close_list(&mut out, &mut in_list);
        para.push(trimmed.to_owned());
    }
    close_list(&mut out, &mut in_list);
    close_para(&mut out, &mut para);
    out.join("\n")
}

fn parse_heading(trimmed: &str) -> Option<String> {
    let hashes = trimmed.chars().take_while(|ch| *ch == '#').count();
    if !(1..=3).contains(&hashes) {
        return None;
    }
    let rest = trimmed[hashes..].strip_prefix(' ')?;
    let level = (hashes + 1).min(3);
    Some(format!("<h{level}>{}</h{level}>", inline(rest)))
}

fn parse_banner(trimmed: &str) -> Option<String> {
    let inner = trimmed.strip_prefix("---")?.strip_suffix("---")?.trim();
    if inner.is_empty() || inner.contains("---") {
        return None;
    }
    Some(format!("<h2>{}</h2>", inline(inner)))
}

fn parse_item(trimmed: &str) -> Option<&str> {
    trimmed
        .strip_prefix('-')
        .or_else(|| trimmed.strip_prefix('*'))
        .and_then(|rest| rest.strip_prefix(' '))
        .map(str::trim)
}

fn inline(text: &str) -> String {
    let escaped = crate::i18n::escape_html(text);
    apply_markup(&escaped, "**", "strong", "em")
}

fn apply_markup(text: &str, token: &str, strong: &str, em: &str) -> String {
    let mut out = String::with_capacity(text.len());
    let mut rest = text;
    while let Some(start) = rest.find(token) {
        let after = &rest[start + token.len()..];
        if let Some(end) = after.find(token) {
            out.push_str(&rest[..start]);
            out.push_str(&format!("<{strong}>{}</{strong}>", &after[..end]));
            rest = &after[end + token.len()..];
        } else {
            break;
        }
    }
    out.push_str(rest);
    if token == "**" {
        return apply_markup(&out, "*", em, em);
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn renders_headings_lists_and_markup() {
        let html = render_txt_to_html("# Title\n\n- one\n- **two**\n\nplain *em* text");
        assert!(html.contains("<h2>Title</h2>"));
        assert!(html.contains("<ul>"));
        assert!(html.contains("<li><strong>two</strong></li>"));
        assert!(html.contains("<p>plain <em>em</em> text</p>"));
    }

    #[test]
    fn legal_texts_render_nonempty() {
        assert!(!render_txt_to_html(TERMS_TXT).is_empty());
        assert!(!render_txt_to_html(PRIVACY_TXT).is_empty());
    }
}
